src/serialization/src/main/kotlin/org/vaccineimpact/api/serialization/ModelBinder.kt

```diff
@@ -3,53 +3,75 @@ package org.vaccineimpact.api.serialization
 import org.vaccineimpact.api.models.ErrorInfo
 import org.vaccineimpact.api.models.validation.CanBeBlank
 import org.vaccineimpact.api.serialization.validation.ValidationException
-import org.vaccineimpact.api.serialization.validation.applyRule
-import org.vaccineimpact.api.serialization.validation.checkMissing
+import org.vaccineimpact.api.serialization.validation.Validator
 import kotlin.reflect.KClass
 import kotlin.reflect.KProperty1
 import kotlin.reflect.full.findAnnotation
 import kotlin.reflect.full.memberProperties
 
-class ModelBinder(private val serializer: Serializer = MontaguSerializer.instance)
+class ModelBinder(private val serializer: Serializer = MontaguSerializer.instance,
+                  private val validator: Validator = Validator(serializer))
 {
     fun <T : Any> deserialize(body: String, klass: Class<T>): T
     {
         val model = serializer.fromJson(preprocessed(body), klass)
         val errors = verify(model)
         if (errors.any())
         {
-            throw ValidationException(errors)
+            throw ValidationException(errors.distinct())
         }
         return model
     }
 
     fun verify(model: Any): List<ErrorInfo>
     {
         val properties = model::class.memberProperties.filterIsInstance<KProperty1<Any, *>>()
-        return properties.flatMap { verify(it, model) }
+        return properties.flatMap { recursiveVerify(it, model) ?: listOf() }
     }
 
-    fun verify(property: KProperty1<Any, *>, model: Any): List<ErrorInfo>
+    private fun recursiveVerify(property: KProperty1<Any, *>, model: Any): List<ErrorInfo>
+    {
+        val errors = verify(property, model)
+
+        // end condition
+        // value is null
+        val value = property.get(model) ?:
+                return errors
+
+        val klass = value::class
+        if (!klass.isData)
+        {
+            // end condition
+            // primitive property
+            return errors
+        }
+
+        val members = klass.memberProperties.filterIsInstance<KProperty1<Any, *>>()
+        return errors + members.flatMap { recursiveVerify(it, value) }
+    }
```

Comment:
Maybe clearer just to write this with a mutable errors? Then you can have single point of return. i.e.

```
var errors = verify(property, model)
val value = property.get(model)
if (value != null && value::class.isData)
{
    val members = value::class.memberProperties.filterIsInstance<KProperty1<Any, *>>()
    errors += members.flatMap { recursiveVerify(it, value) }
}
return errors
```